design-documents/buildattr64-rationale.rst

```diff
@@ -0,0 +1,545 @@
+  from an optional subsection at once if they don't recognize the
+  subsection name.
+
+- Subsections may share common implementation properties such as all
+  tags having the same partial order which could make them easier to
+  process. For example if all tags in a subsection have the value 0
+  or 1, and have a partial order that is identical or reverse of the
+  tags arithmetic value, then an implementation can represent this as
+  a bit vector and use the binary ``or`` or ``and`` operator to do the
+  join.
+
+Alternative: one .aeabi_attributes section
+------------------------------------------
+
+Instead of grouping attributes with similar properties into their own
+subsection, we follow (ADDENDA32_) and have one single public
+.aeabi_attributes section that contains all the public attributes. Or
+have one public .aeabi_attributes section for ULEB128 and one for
+NTBS.
+
+This would be simpler to encode in assembly, with only one directive
+required, but would make the specification harder to understand
+and maintain consistency over time.
+
+All tags ULEB128 or NTBS
+========================
+
+A build attributes consumer can more easily skip tags it doesn't
+understand if the encoding for the value is known ahead of time. The
+rule in (ADDENDA32_) that used whether the tag was even or odd was
+error prone as the rule is not widely known. Moreover almost all
+tags are ULEB128 so the even odd rule wastes almost half the encoding
+space for tags.
+
+Reference Partial Order
+=======================
+
+A majority of build attributes have the partial order either matching
+or reversing the build attribute arithmetic order. As (ADDENDA32_)
+states it is usually obvious when this is the case. However there are
+more complex cases and in these cases it can help to state what the
+reference intention is for the tag.
+
+Appendix: Recommended Tool Interface for aeabi subsections
+==========================================================
+
+Compiler
+--------
+
+Build attributes are set by the compiler based on command-line
+options.  For example the clang and gcc ``-mbranch-protection`` option
+can be used to derive ``Tag_Feature_BTI`` and ``Tag_Feature_PAC`` and
+``Tag_Feature_GCS``.
+
+Individual functions can be given different values from the file scope
+command-line options. The file scope build attributes should still be
+derived from the file scope command-line options, or module level encodings
+of the file scope command-line options in the case of link time optimization.
+It is the user's responsibility that the individual functions are used in a
+compatible way to the file scope build attributes.
+
+Assembler
+---------
+
+Where possible the assembler can derive build attributes from the
+assembler's command line options in the same way as the compiler.  For
+options that cannot be derived the following directives can be used to
+construct "aeabi" prefixed subsections.
+
+::
+   .aeabi_subsection name [optional] [, parameter type]
```

Comment:
The syntax is not final. There are consideration around whether the values should be optional and with what exact syntax.
Options:
- as it is now, and to denote optional arguments: 
 .aeabi_subsection MySectionA 0 -> 'optional'
 .aeabi_subsection MySectionA, 0 -> 'parameter type'
- The syntax might be .aeabi_subsection name [, optional] [, parameter type]
and to denote optional arguments:
.aeabi_subsection MySectionA, 0,   or: 
.aeabi_subsection MySectionA, ,0
- Having both the arguments being mandatory
- Make encoding and parameter type either both present or absent

Reply:
Internally with a conversation with the GNU team we decided that it should be `[, optional]` I think that while possible to disambiguate between
```
.aeabi-subsection vendor-name optional
```
and
```
.aeabi-subsection vendor-name ,ntbs
```
Is possible, given that optional and ,ntbs can be digits then it will be difficult for a human to spot.